oneview/resource_server_profile_template.go

```diff
@@ -1198,38 +1219,47 @@ func resourceServerProfileTemplateCreate(d *schema.ResourceData, meta interface{
 		}
 		serverProfileTemplate.SanStorage.VolumeAttachments = volumeAttachments
 	}
-	rawOsDeploySetting := d.Get("os_deployment_settings").(*schema.Set).List()
-	osDeploySetting := ov.OSDeploymentSettings{}
-	for _, raw := range rawOsDeploySetting {
-		osDeploySettingItem := raw.(map[string]interface{})
-		osDeploymentPlan, err := config.ovClient.GetOSDeploymentPlanByName(osDeploySettingItem["os_deployment_plan_name"].(string))
-		if err != nil {
-			return err
-		}
-		if osDeploymentPlan.URI == "" {
-			return fmt.Errorf("Could not find deployment plan by name: %s", osDeploySettingItem["os_deployment_plan_name"].(string))
-		}
 
-		osCustomAttributes := make([]ov.OSCustomAttribute, 0)
-		if osDeploySettingItem["os_custom_attributes"] != nil {
-			rawOsDeploySettings := osDeploySettingItem["os_custom_attributes"].(*schema.Set).List()
-			for _, rawDeploySetting := range rawOsDeploySettings {
-				rawOsDeploySetting := rawDeploySetting.(map[string]interface{})
-				osCustomAttributes = append(osCustomAttributes, ov.OSCustomAttribute{
-					Name:  rawOsDeploySetting["name"].(string),
-					Value: rawOsDeploySetting["value"].(string),
-				})
+	if _, ok := d.GetOk("os_deployment_settings"); ok {
+		rawOsDeploySetting := d.Get("os_deployment_settings").(*schema.Set).List()
+		osDeploySetting := ov.OSDeploymentSettings{}
+		for _, raw := range rawOsDeploySetting {
+			osDeploySettingItem := raw.(map[string]interface{})
+			osdp := ""
+			if osDeploySettingItem["os_deployment_plan_uri"] != "" {
+				osdp = osDeploySettingItem["os_deployment_plan_uri"].(string)
+			} else if osDeploySettingItem["os_deployment_plan_name"] != "" {
+				osDeploymentPlan, err := config.ovClient.GetOSDeploymentPlanByName(osDeploySettingItem["os_deployment_plan_name"].(string))
+				if err != nil {
+					return err
+				}
+				if osDeploymentPlan.URI == "" {
+					return fmt.Errorf("Could not find deployment plan by name: %s", osDeploySettingItem["os_deployment_plan_name"].(string))
+				}
+				osdp = osDeploymentPlan.URI.String()
 			}
-		}
 
-		osDeploySetting = ov.OSDeploymentSettings{
-			OSDeploymentPlanUri: osDeploymentPlan.URI,
-			OSVolumeUri:         utils.NewNstring(osDeploySettingItem["os_volume_uri"].(string)),
-			OSCustomAttributes:  osCustomAttributes,
+			osCustomAttributes := make([]ov.OSCustomAttribute, 0)
+			if osDeploySettingItem["os_custom_attributes"] != nil {
+				rawOsDeploySettings := osDeploySettingItem["os_custom_attributes"].(*schema.Set).List()
+				for _, rawDeploySetting := range rawOsDeploySettings {
+					rawOsDeploySetting := rawDeploySetting.(map[string]interface{})
+					osCustomAttributes = append(osCustomAttributes, ov.OSCustomAttribute{
+						Name:  rawOsDeploySetting["name"].(string),
+						Value: rawOsDeploySetting["value"].(string),
```

Comment:
if constraints and type fields are available, we can set them as well since they are defined in schema.

Reply:
we don't pass them in create calls. They gets created by Oneview. So when those parameters are not provided by user, mapping is not required.

Reply:
In that case, change to Computed instead of Optional.